File: archive/cses/geometry/2190-line-segment-intersection/main.cc
#include <bits/stdc++.h>

using namespace std;

template <typename T>
int sgn(T val) {
    if (val < 0) return -1;
    if (val > 0) return 1;
    return 0;
}

template <typename T>
struct Point {
    using P=Point;
    T x, y;

    Point(T _x = 0, T _y = 0) : x(_x), y(_y) {}
    P operator+(P o) const { return P(x + o.x, y + o.y); }
    P operator-(P o) const { return P(x - o.x, y - o.y); }
    P operator*(T d) const { return P(x * d, y * d); }
    P operator/(T d) const { return P(x / d, y / d); }
    T dot(P o) const { return x * o.x + y * o.y; }
    T cross(P o) const { return x * o.y - y * o.x; }
    T abs2() const { return x * x + y * y; }
    long double abs() const { return sqrt((long double) abs2()); }
    double angle() const { return atan2(y, x); } // [-pi, pi]
    P unit() const { return *this / abs(); } // makes abs()=1
    P perp() const { return P(-y, x); } // rotates +pi/2

    P rotate(double a) const { // ccw
        return P(x * cos(a) - y * sin(a), x * sin(a) + y * cos(a));
    }

    friend ostream& operator<<(ostream& os, P p) {
        return os << "(" << p.x << ", " << p.y << ")";
    }
};

using P = Point<int64_t>;

bool colinear(const P& v1, const P& v2) {
    return v1.cross(v2) == 0;
}

bool checkBoundingBox(
        const P& p1,
        const P& q1,
        const P& p2,
        const P& q2
) {
    if (max(p1.x, q1.x) < min(p2.x, q2.x)) return true;
    if (max(p1.y, q1.y) < min(p2.y, q2.y)) return true;
    if (max(p2.x, q2.x) < min(p1.x, q1.x)) return true;
    if (max(p2.x, q2.x) < min(p1.x, q1.x)) return true;
    return false;
}

bool intersect(P p1, P p2, P p3, P p4) {
    // Check if two segments are parallel
    if (colinear(p2 - p1, p4 - p3)) {
        // Check if 4 points are colinear
        if (!colinear(p2 - p1, p3 - p1)) return false;
        if (checkBoundingBox(p1, p2, p3, p4)) return false;
        return true;
    }

    // check if one line is completely on one side of the other
    for (int i = 0; i < 2; i++) {
        if (sgn((p2 - p1).cross(p3 - p1)) == sgn((p2 - p1).cross(p4 - p1))
                && sgn((p2 - p1).cross(p3 - p1)) != 0) {
            return false;
        }
        swap(p1, p3);
        swap(p2, p4);
    }
    return true;
}

int main() {
    int t;
    cin >> t;

    for (int ti = 0; ti < t; ti++) {
        vector<P> p;
        for (int i = 0; i < 4; i++) {
            int x, y;
            cin >> x >> y;
            p.emplace_back(x, y);
        }
        P p1 = p[0];
        P p2 = p[1];
        P p3 = p[2];
        P p4 = p[3];
        cout << (solve(p1, p2, p3, p4) ? "YES" : "NO") << '\n';
    }

    return 0;
}

File: archive/cses/geometry/2190-line-segment-intersection/Main.java
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.StringTokenizer;

public class Main {
  public static void main(String[] args) {
    InputReader in = new InputReader(System.in);
    PrintWriter out = new PrintWriter(System.out);
    int t = in.nextInt();
    for (int i = 0; i < t; i++) {
      solve(i, in, out);
    }
    out.close();
  }

  public static void solve(int caseId, InputReader in, PrintWriter out) {
    Segment s = new Segment(
        new Point(in.nextInt(), in.nextInt()),
        new Point(in.nextInt(), in.nextInt())
        );
    Segment r = new Segment(
        new Point(in.nextInt(), in.nextInt()),
        new Point(in.nextInt(), in.nextInt())
        );
    out.println(Geometry.intersecting(s, r) ? "YES" : "NO");
  }

  static class Geometry {
    static double EPS = 1e-9;

    static int signum(final double x) {
      if (Math.abs(x) < EPS) return 0;
      return Double.compare(x, 0);
    }

    // Check the orientation of three points: p1, p2, p3.
    // If > 0: p3 is on the left of the vector p1p2.
    // If < 0: p3 is on the right of the vector p1p2.
    static double orientation(final Point p1, final Point p2, final Point p3) {
      return orientation(p2.sub(p1), p3.sub(p1));
    }

    // Check the orientation of two vectors: v and w;
    // If > 0: v is on the right of w.
    // If < 0: v is on the left of w.
    // Way to remember: Given the two unit vectors i, j in 2D space.
    // We always have: i.cross(j) = 1.
    static double orientation(final Point v1, final Point v2) {
      return v1.cross(v2);
    }

    static boolean parallel(final Segment s, final Segment r) {
      double cross = s.asVec().cross(r.asVec());
      return signum(cross) == 0;
    }

    static boolean intersecting(Segment s, Segment r) {
      if (parallel(s, r)) {
        // parallel
        if (signum(orientation(s.p, s.q, r.p)) != 0) {
          // but not colinear
          return false;
        } else {
          // colinear. we check the bounding boxes
          for (int i = 0; i < 2; i++) {
            if (s.xMax() < r.xMin() || s.yMax() < r.yMin()) {
              return false;
            }
            Segment t = s;
            s = r;
            r = t;
          }
          return true;
        }
      } else {
        for (int i = 0; i < 2; i++) {
          double rpRelToS = orientation(s.p, s.q, r.p);
          double rqRelToS = orientation(s.p, s.q, r.q);
          // Check if rp and rq are (completely) on the same side relative to s.
          if (signum(rpRelToS) == signum(rqRelToS) && signum(rpRelToS) != 0) {
            return false;
          }
          Segment t = s;
          s = r;
          r = t;
        }
        return true;
      }
    }
  }

  static class Segment {
    Point p;
    Point q;

    Segment(Point p, Point q) {
      this.p = p;
      this.q = q;
    }

    Point asVec() {
      return q.sub(p);
    }

    double xMin() {
      return Math.min(p.x, q.x);
    }

    double xMax() {
      return Math.max(p.x, q.x);
    }

    double yMin() {
      return Math.min(p.y, q.y);
    }

    double yMax() {
      return Math.max(p.y, q.y);
    }
  }

  static class Point {
    public double x;
    public double y;

    public Point(double x, double y) {
      this.x = x;
      this.y = y;
    }

    public Point clone() {
      return new Point(x, y);
    }

    public Point addSelf(final double x, final double y) {
      this.x += x;
      this.y += y;
      return this;
    }

    public Point addSelf(final Point o) {
      return addSelf(o.x, o.y);
    }

    public Point subSelf(final double x, final double y) {
      this.x -= x;
      this.y -= y;
      return this;
    }

    public Point subSelf(final Point o) {
      return subSelf(o.x, o.y);
    }

    public Point add(final Point o) {
      return new Point(x + o.x, y + o.y);
    }

    public Point sub(final Point o) {
      return new Point(x - o.x, y - o.y);
    }

    public double dot(final Point o) {
      return x * o.x + y * o.y;
    }

    public double cross(final Point o) {
      return x * o.y - y * o.x;
    }
  }

  static class InputReader {
    private final BufferedReader reader;
    private StringTokenizer tokenizer;

    public InputReader(InputStream inputStream) {
      reader = new BufferedReader(new InputStreamReader(inputStream), 32768);
      tokenizer = new StringTokenizer("");
    }

    public boolean hasNext() {
      if (!tokenizer.hasMoreTokens()) {
        String line = null;
        try {
          line = reader.readLine();
        } catch (IOException e) {
          e.printStackTrace();
        }
        if (line == null) {
          return false;
        }
        tokenizer = new StringTokenizer(line);
      }
      return true;
    }

    public String next() {
      if (!hasNext()) {
        throw new RuntimeException();
      }
      return tokenizer.nextToken();
    }

    public int nextInt() {
      return Integer.parseInt(next());
    }

    public long nextLong() {
      return Long.parseLong(next());
    }

    public double nextDouble() {
      return Double.parseDouble(next());
    }
  }
}

